docs/user/settings.rst
Settings
========

.. include:: /partials/settings-note.rst

.. _openwisp_notifications_web_enabled:

``OPENWISP_NOTIFICATIONS_WEB_ENABLED``
--------------------------------------

======= ========
Type    ``bool``
Default ``True``
======= ========

This setting controls the global default for web notifications across the
entire system. When set to ``False``, web notifications will be disabled
by default for all news users and organizations.

This setting can be overridden at both :doc:`organization and user levels
<notification-preferences>`. Organization administrators can configure
different defaults for their organization, and individual users can set
their own preferences regardless of the global or organization settings.

.. _openwisp_notifications_email_enabled:

``OPENWISP_NOTIFICATIONS_EMAIL_ENABLED``
----------------------------------------

======= ========
Type    ``bool``
Default ``True``
======= ========

This setting controls the global default for email notifications across
the entire system. When set to ``False``, email notifications will be
disabled by default for all new users and organizations.

This setting can be overridden at both :doc:`organization and user levels
<notification-preferences>`. Organization administrators can configure
different defaults for their organization, and individual users can set
their own preferences regardless of the global or organization settings.

.. note::

    These global settings serve as the foundation for the notification
    system. For more information about how users and organizations can
    customize their notification preferences, see
    :doc:`notification-preferences`.

.. _openwisp_notifications_host:

``OPENWISP_NOTIFICATIONS_HOST``
-------------------------------

======= ======================================
type    ``str``
default Any domain defined in ``ALLOWED_HOST``
======= ======================================

This setting defines the domain at which API and Web Socket communicate
for working of notification widget.

.. note::

    You don't need to configure this setting if you don't host your API
    endpoints on a different sub-domain.

If your root domain is ``example.com`` and API and Web Socket are hosted
at ``api.example.com``, then configure setting as follows:

.. code-block:: python

    OPENWISP_NOTIFICATIONS_HOST = "https://api.example.com"

This feature requires you to allow `CORS
<https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS>`_ on your server.
We use ``django-cors-headers`` module to easily setup CORS headers. Please
refer `django-core-headers' setup documentation
<https://github.com/adamchainz/django-cors-headers#setup>`_.

Configure ``django-cors-headers`` settings as follows:

.. code-block:: python

    CORS_ALLOW_CREDENTIALS = True
    CORS_ORIGIN_WHITELIST = ["https://www.example.com"]

Configure Django's settings as follows:

.. code-block:: python

    SESSION_COOKIE_DOMAIN = "example.com"
    CSRF_COOKIE_DOMAIN = "example.com"

Please refer to `Django's settings documentation
<https://docs.djangoproject.com/en/4.2/ref/settings/>`_ for more
information on ``SESSION_COOKIE_DOMAIN`` and ``CSRF_COOKIE_DOMAIN``
settings.

``OPENWISP_NOTIFICATIONS_SOUND``
--------------------------------

======= ===================================================================================================================================================
type    ``str``
default `notification_bell.mp3
        <https://github.com/openwisp/openwisp-notifications/tree/master/openwisp_notifications/static/openwisp-notifications/audio/notification_bell.mp3>`_
======= ===================================================================================================================================================

This setting defines notification sound to be played when notification is
received in real-time on admin site.

Provide a relative path (hosted on your web server) to audio file as show
below.

.. code-block:: python

    OPENWISP_NOTIFICATIONS_SOUND = "your-appname/audio/notification.mp3"

.. _openwisp_notifications_cache_timeout:

``OPENWISP_NOTIFICATIONS_CACHE_TIMEOUT``
----------------------------------------

======= =================================
type    ``int``
default ``172800`` `(2 days, in seconds)`
======= =================================

It sets the number of seconds the notification contents should be stored
in the cache. If you want cached notification content to never expire,
then set it to ``None``. Set it to ``0`` if you don't want to store
notification contents in cache at all.

.. _openwisp_notifications_ignore_enabled_admin:

``OPENWISP_NOTIFICATIONS_IGNORE_ENABLED_ADMIN``
-----------------------------------------------

======= ========
type    ``list``
default []
======= ========

This setting enables the widget which allows users to :ref:`silence
notifications for specific objects temporarily or permanently.
<notifications_silencing>` in the change page of the specified
``ModelAdmin`` classes.

E.g., if you want to enable the widget for objects of
``openwisp_users.models.User`` model, then configure the setting as
following:

.. code-block:: python

    OPENWISP_NOTIFICATIONS_IGNORE_ENABLED_ADMIN = ["openwisp_users.admin.UserAdmin"]

``OPENWISP_NOTIFICATIONS_POPULATE_PREFERENCES_ON_MIGRATE``
----------------------------------------------------------

======= ========
type    ``bool``
default ``True``
======= ========

This setting allows to disable creating :doc:`notification preferences
<notification-preferences>` on running migrations.

.. _openwisp_notifications_notification_storm_prevention:

``OPENWISP_NOTIFICATIONS_NOTIFICATION_STORM_PREVENTION``
--------------------------------------------------------

When the system starts creating a lot of notifications because of a
general network outage (e.g.: a power outage, a global misconfiguration),
the notification storm prevention mechanism avoids the constant displaying
of new notification alerts as well as their sound, only the notification
counter will continue updating periodically, although it won't emit any
sound or create any other visual element until the notification storm is
over.

This setting allows tweaking how this mechanism works.

For details on how the storm prevention affects real-time WebSocket
delivery, refer to the :doc:`WebSocket API reference <websocket-api>`.

The default configuration is as follows:

.. code-block:: python

    OPENWISP_NOTIFICATIONS_NOTIFICATION_STORM_PREVENTION = {
        # Time period for tracking burst of notifications (in seconds)
        "short_term_time_period": 10,
        # Number of notifications considered as a notification burst
        "short_term_notification_count": 6,
        # Time period for tracking notifications in long time interval (in seconds)
        "long_term_time_period": 180,
        # Number of notifications in long time interval to be considered as a notification storm
        "long_term_notification_count": 30,
        # Initial time for which notification updates should be skipped (in seconds)
        "initial_backoff": 1,
        # Time by which skipping of notification updates should be increased (in seconds)
        "backoff_increment": 1,
        # Maximum interval after which the notification widget should get updated (in seconds)
        "max_allowed_backoff": 15,
    }

.. _openwisp_notifications_email_batch_interval:

``OPENWISP_NOTIFICATIONS_EMAIL_BATCH_INTERVAL``
-----------------------------------------------

======= ===============================
Type    ``int``
Default ``10800`` (3 hours, in seconds)
======= ===============================

This setting determines the :ref:`interval of the email batching feature
<notifications_batches>`.

The interval is specified in seconds.

To send email notifications immediately without batching, set this value
to ``0``.

.. _openwisp_notifications_email_batch_display_limit:

``OPENWISP_NOTIFICATIONS_EMAIL_BATCH_DISPLAY_LIMIT``
----------------------------------------------------

======= =======
Type    ``int``
Default ``15``
======= =======

This setting specifies the maximum number of email notifications that can
be included in a single :ref:`email batch <notifications_batches>`.


CHANGES.rst
Changelog
=========

Version 1.3.0 [unreleased]
--------------------------

Work in progress.

Version 1.2.3 [2026-06-04]
--------------------------

Bugfixes
~~~~~~~~

- Filtered out soft-deleted notification settings in preferences API `#452
  <https://github.com/openwisp/openwisp-notifications/issues/452>`_
- Fixed accidental email delivery for generic_message notifications `#478
  <https://github.com/openwisp/openwisp-notifications/issues/478>`_
- Restrict notification preferences page to staff or super users `#462
  <https://github.com/openwisp/openwisp-notifications/issues/462>`_
- Create global notification settings for non-org admin staff users `#453
  <https://github.com/openwisp/openwisp-notifications/issues/453>`_

Version 1.2.2 [2026-05-12]
--------------------------

Bugfixes
~~~~~~~~

- Fixed multiple bugs affecting notification preferences `#448
  <https://github.com/openwisp/openwisp-notifications/issues/448>`_:

  1. Creating a new organization ignored the user's global notification
     preferences.
  2. Promoting or adding an organization admin did not respect global
     preferences.
  3. Changes to organization-level settings could overwrite user-specific
     preferences.

Version 1.2.1 [2025-11-27]
--------------------------

Bugfixes
~~~~~~~~

- Fixed accidental overwriting of user preferences `#427
  <https://github.com/openwisp/openwisp-notifications/issues/427>`_

Version 1.2.0 [2025-10-23]
--------------------------

Features
~~~~~~~~

- Added `dedicated notification preferences page
  <https://openwisp.io/docs/dev/notifications/user/notification-preferences.html>`_.
- Added `organization notification settings
  <https://openwisp.io/docs/dev/notifications/user/notification-preferences.html#organization-settings>`_.
- Added `Email Batch Summary
  <https://openwisp.io/docs/dev/notifications/user/web-email-notifications.html#email-batches>`_.
- Added email verification warning `#281
  <https://github.com/openwisp/openwisp-notifications/issues/281>`_.

Changes
~~~~~~~

Other changes
+++++++++++++

- Improved UI of email notifications.
- Disallowed changing preferences for ``generic_message`` notification
  type `#359
  <https://github.com/openwisp/openwisp-notifications/issues/359>`_.
- Dropped dependency on ``django-notifications`` `#353
  <https://github.com/openwisp/openwisp-notifications/issues/353>`_.
- Removed deprecated code for updating superuser notification settings
  `#283 <https://github.com/openwisp/openwisp-notifications/issues/283>`_.

Dependencies
++++++++++++

- Bumped ``openwisp-users~=1.2.0``.
- Bumped ``openwisp-utils~=1.2.0``.
- Bumped ``markdown~=3.9``.
- Added support for Django ``5.x``.
- Added support for Python ``3.11``, ``3.12``, and ``3.13``.
- Dropped support for Django ``3.2`` and ``4.1``.
- Dropped support for Python ``3.8``.

Bugfixes
~~~~~~~~

- Prevented notification toast from appearing while leaving a page `#264
  <https://github.com/openwisp/openwisp-notifications/issues/264>`_.
- Ensured notification widget closes only on user-initiated clicks.
- Fixed notification widget loading on non-admin pages `#271
  <https://github.com/openwisp/openwisp-notifications/issues/271>`_.
- Allowed overriding ``openwisp_users`` app label in tests.
- Hid notification preferences for disabled organizations `#366
  <https://github.com/openwisp/openwisp-notifications/issues/366>`_.
- Fixed JavaScript logic for creating notification toasts.
- Fixed issue with notification widget removing URL fragments.

Version 1.1.1 [2025-01-31]
--------------------------

Bugfixes
~~~~~~~~

- Fixed `unnecessary triggering of
  "update_superuser_notification_settings" celery task
  <https://github.com/openwisp/openwisp-notifications/issues/283>`_.

Version 1.1.0 [2024-11-20]
--------------------------

Features
~~~~~~~~

- Added support for `defining related object links in notification
  configurations
  <https://github.com/openwisp/openwisp-notifications/pull/250>`_.
- Introduced a `"generic_message" notification type
  <https://openwisp.io/docs/stable/notifications/user/notification-types.html#generic-message>`_
  to deliver custom notifications in the user interface.
- Enabled filtering of ``NotificationSetting`` using ``organization_slug``
  in the REST API.

Changes
~~~~~~~

- `Only users with verified email addresses will receive email
  notifications
  <https://github.com/openwisp/openwisp-notifications/issues/270>`_.

Dependencies
++++++++++++

- Bumped ``django-notifications-hq~=1.8.3``.
- Bumped ``markdown~=3.6.0``.
- Bumped ``openwisp-users~=1.1.0``.
- Bumped ``openwisp-utils[rest,celery]~=1.1.1``.
- Added support for Python ``3.10``.
- Dropped support for Python ``3.7``.
- Added support for Django ``4.2.x``.
- Dropped support for Django ``4.0.x``.

Bugfixes
~~~~~~~~

- Fixed `issue where users received multiple identical notifications
  <https://github.com/openwisp/openwisp-notifications/issues/277>`_ when
  member of multiple organizations.
- Increased ``timeoutInterval`` for the ``ReconnectingWebSocket`` to ``7``
  seconds to prevent termination due to timeout on slow network
  connections.

Version 1.0.3 [2022-08-03]
--------------------------

Bugfixes
~~~~~~~~

- Flagged tests that should not be run on a production environment: These
  tests depend on the static storage backend of the project. In a
  production environment, the filenames could get changed due to static
  minification and cache invalidation. Hence, these tests should not be
  run on the production environment because they'll fail.

Version 1.0.2 [2022-07-01]
--------------------------

Bugfixes
~~~~~~~~

- Fixed `hardcoded static image URLs
  <https://github.com/openwisp/openwisp-notifications/issues/243>`_. These
  create issues when static files are served using an external service
  (e.g. S3 storage buckets).
- Fixed `"Organization.DoesNotExist" error on creating a new organization
  <https://github.com/openwisp/openwisp-notifications/issues/238>`_.

Version 1.0.1 [2022-06-09]

openwisp_notifications/settings.py
import re

from django.conf import settings

CONFIG_DEFAULTS = {
    "PAGINATE_BY": 20,
    "USE_JSONFIELD": True,
    "SOFT_DELETE": False,
    "NUM_TO_FETCH": 10,
    "CACHE_TIMEOUT": 2,
}
DISALLOW_PREFERENCES_CHANGE_TYPE = ["generic_message"]


def get_setting(name, default=None):
    """
    Get a setting from the Django settings module or return a default value.
    """
    return getattr(settings, f"OPENWISP_NOTIFICATIONS_{name}", default)


HOST = get_setting("HOST", None)

CACHE_TIMEOUT = get_setting("CACHE_TIMEOUT", 2 * 24 * 60 * 60)  # 2 days

IGNORE_ENABLED_ADMIN = get_setting("IGNORE_ENABLED_ADMIN", [])
POPULATE_PREFERENCES_ON_MIGRATE = get_setting("POPULATE_PREFERENCES_ON_MIGRATE", True)
NOTIFICATION_STORM_PREVENTION = get_setting(
    "NOTIFICATION_STORM_PREVENTION",
    {
        "short_term_time_period": 10,
        "short_term_notification_count": 6,
        "long_term_time_period": 180,
        "long_term_notification_count": 30,
        "initial_backoff": 1,
        "backoff_increment": 1,
        "max_allowed_backoff": 15,
    },
)

SOUND = get_setting("SOUND", "openwisp-notifications/audio/notification_bell.mp3")

WEB_ENABLED = get_setting("WEB_ENABLED", True)
EMAIL_ENABLED = get_setting("EMAIL_ENABLED", True)
EMAIL_BATCH_INTERVAL = get_setting("EMAIL_BATCH_INTERVAL", 180 * 60)  # 3 hours
EMAIL_BATCH_DISPLAY_LIMIT = get_setting("EMAIL_BATCH_DISPLAY_LIMIT", 15)


# Remove the leading "/static/" here as it will
# conflict with the "static()" call in context_processors.py.
# This is done for backward compatibility.
SOUND = re.sub("^/static/", "", SOUND)


def get_config():
    user_config = get_setting("CONFIG", {})
    config = CONFIG_DEFAULTS.copy()
    config.update(user_config)
    return config
